File: frontend/src/interfaces/pack.ts
import {Type} from "class-transformer";
import type {LoaderVersions} from "@/stores/cache.ts";
import type {User} from "@/interfaces/user.ts";


export class AllPacksResponse {
  @Type(() => PackResponse)
  packs?: PackResponse[];
}

export class Pack {
  slug!: string;
  name!: string;
  description!: string;
  createdAt!: string;
  createdBy!: number;
  author!: User;
  updatedAt!: string;
  updatedBy!: number;
  deletedAt?: string;
  isPublic!: boolean;
  status!: PackStatus;
  mcVersion!: string;
  loader!: keyof LoaderVersions;
  loaderVersion!: string;
  acceptableGameVersions?: string[];
  version!: string;
  packFormat!: string;
  mods?: Mod[]

  get isArchived(): boolean {
    return this.deletedAt !== null && this.deletedAt !== undefined;
  }
}

export class PackResponse extends Pack {
  currentUserPermission!: PackPermission;
}

export enum PackStatus {
  PUBLISHED = 'published',
  DRAFT = 'draft',
}

export enum PackPermission {
  STATIC = 1,
  VIEW = 10,
  EDIT = 20,
}

export class Mod {
  id!: number;
  packSlug!: string;
  slug!: string;
  name!: string;
  type!: string;
  fileName!: string;
  side!: "client" | "server" | "both";
  pinned!: boolean;
  source!: string;
  update!: {[key: string]: string | number}
  createdBy!: number;
  createdAt!: string;
  updatedBy!: number;
  updatedAt!: string;
}


export class ModDependency {
  fileName!: string;
  modType!: string;
  name!: string;
  side!: "client" | "server" | "both";
  slug!: string;
  url!: string;
}


export class ModDependenciesResponse {
  missing!: ModDependency[]
}

File: backend/internal/tables/pack.go
package tables

import (
	"github.com/leocov-dev/packwiz-nxt/core"
	"gorm.io/datatypes"
	"gorm.io/gorm"
	"packwiz-web/internal/types"
	"time"
)

type Pack struct {
	Slug                   string                      `gorm:"primarykey" json:"slug"`
	Name                   string                      `json:"name"`
	CreatedAt              time.Time                   `json:"createdAt"`
	UpdatedAt              time.Time                   `json:"updatedAt"`
	DeletedAt              gorm.DeletedAt              `gorm:"index" json:"deletedAt"`
	Description            string                      `json:"description"`
	CreatedBy              uint                        `json:"createdBy"`
	Author                 User                        `gorm:"foreignKey:CreatedBy" json:"author"`
	UpdatedBy              uint                        `json:"updatedBy"`
	IsPublic               bool                        `json:"isPublic"`
	Status                 types.PackStatus            `gorm:"default:draft" json:"status"`
	MCVersion              string                      `json:"mcVersion"`
	Loader                 string                      `json:"loader"`
	LoaderVersion          string                      `json:"loaderVersion"`
	AcceptableGameVersions datatypes.JSONSlice[string] `json:"acceptableGameVersions"`
	Version                string                      `json:"version"`
	PackFormat             string                      `json:"packFormat"`

	Mods []Mod `gorm:"foreignKey:PackSlug" json:"mods"`
}

func (p Pack) AsMeta() core.Pack {
	mods := map[string]*core.Mod{}

	for _, mod := range p.Mods {
		mods[mod.Slug] = mod.AsMeta()
	}

	return core.Pack{
		Name:        p.Name,
		Author:      p.Author.Username,
		Version:     p.Version,
		Description: p.Description,
		PackFormat:  p.PackFormat,
		Versions: map[string]string{
			"minecraft": p.MCVersion,
			p.Loader:    p.LoaderVersion,
		},
		Export: nil,
		Options: map[string]interface{}{
			"acceptable-game-versions": []string(p.AcceptableGameVersions),
		},
		Mods: mods,
	}
}
